src/groupeddataframe/grouping.jl

```diff
@@ -177,6 +213,292 @@ end
+_groupvalues(gd::GroupedDataFrame) = [_groupvalues(gd, i) for i in 1:length(gd)]
+
+# Get values of grouping columns for single group
+_groupvalues(gd::GroupedDataFrame, i::Integer) = gd.parent[gd.idx[gd.starts[i]], gd.cols]
+
+# Get values of single grouping column for single group
+_groupvalues(gd::GroupedDataFrame, i::Integer, col::Integer) = gd.parent[gd.idx[gd.starts[i]], gd.cols[col]]
+_groupvalues(gd::GroupedDataFrame, i::Integer, col::Symbol) = _groupvalues(gd, i, _groupvar_idx(gd, col, true))
+
+# Eltypes of the grouping columns
+_grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
+
+
+"""
+    GroupKey{T<:GroupedDataFrame}
+
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
+of the corresponding grouping columns and behaves similarly to a `NamedTuple`,
+but using it to index its `GroupedDataFrame` is much more effecient than using the
+equivalent `Tuple` or `NamedTuple`.
+
+Instances of this type are returned by `keys(::GroupedDataFrame)` and are not
+meant to be constructed directly.
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{T<:GroupedDataFrame}
+    parent::T
+    idx::Int
+end
+
+function Base.show(io::IO, k::GroupKey)
+    print(io, "GroupKey: ")
+    show(io, NamedTuple(k))
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.keys(key::GroupKey) = Tuple(groupvars(parent(key)))
+Base.names(key::GroupKey) = groupvars(parent(key))
+# Private fields are never exposed since they can conflict with column names
+Base.propertynames(key::GroupKey, private::Bool=false) = keys(key)
+Base.values(key::GroupKey) = Tuple(_groupvalues(parent(key), getfield(key, :idx)))
+
+Base.iterate(key::GroupKey, i::Integer=1) = i <= length(key) ? (key[i], i + 1) : nothing
+
+Base.getindex(key::GroupKey, i::Integer) = _groupvalues(parent(key), getfield(key, :idx), i)
+
+function Base.getindex(key::GroupKey, n::Symbol)
+    try
+        return _groupvalues(parent(key), getfield(key, :idx), n)
+    catch e
+        throw(KeyError(n))
+    end
+end
+
+function Base.getproperty(key::GroupKey, p::Symbol)
+    try
+        return key[p]
+    catch e
+        throw(ArgumentError("$(typeof(key)) has no property $p"))
+    end
+end
+
+function Base.NamedTuple(key::GroupKey)
+    N = NamedTuple{Tuple(groupvars(parent(key)))}
+    N(_groupvalues(parent(key), getfield(key, :idx)))
+end
+Base.Tuple(key::GroupKey) = values(key)
+
+
+"""
+    GroupKeys{T<:GroupedDataFrame} <: AbstractVector{GroupKey{T}}
+
+A vector containing all [`GroupKey`](@ref) objects for a given
+[`GroupedDataFrame`](@ref).
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKeys{T<:GroupedDataFrame} <: AbstractVector{GroupKey{T}}
+    parent::T
+end
+
+Base.parent(gk::GroupKeys) = gk.parent
+
+Base.size(gk::GroupKeys) = (length(parent(gk)),)
+Base.IndexStyle(::Type{GroupKeys}) = IndexLinear()
```

Comment:
You didn't apply my suggestion exactly. The `<:` is really needed.
```suggestion
Base.IndexStyle(::Type{<:GroupKeys}) = IndexLinear()
```
(BTW, please don't squash commits as it makes it painful to spot what changed.)

Reply:
Why is it needed? `GroupKeys` is a `DataType`/leaf type so `<:GroupKeys` shouldn't match anything but `GroupKeys` itself, right?

Reply:
This is due a standard problem with parametric types, e.g.:
```
julia> Vector{Int} isa Type{Vector}
false

julia> Vector{Int} isa Type{<:Vector}
true
```

Reply:
Aha, you're absolutely right. Fixed.